lib/Dialect/Arc/Transforms/LowerState.cpp

```diff
@@ -619,70 +626,77 @@ LogicalResult ModuleLowering::lowerExtModule(InstanceOp instOp) {
     baseName.resize(baseNameLen);
     baseName += '/';
     baseName += cast<StringAttr>(name).getValue();
-    auto state = builder.create<AllocStateOp>(
+    auto state = stateBuilder.create<AllocStateOp>(
         result.getLoc(), StateType::get(intType), storageArg);
-    state->setAttr("name", builder.getStringAttr(baseName));
+    state->setAttr("name", stateBuilder.getStringAttr(baseName));
     replaceValueWithStateRead(result, state);
   }
 
-  if (&*builder.getInsertionPoint() == instOp)
-    builder.setInsertionPointAfter(instOp);
   instOp.erase();
   return success();
 }
 
 LogicalResult ModuleLowering::cleanup() {
+  // Clean up dead ops in the model.
+  SetVector<Operation *> erasureWorklist;
+  auto isDead = [](Operation *op) {
+    if (isOpTriviallyDead(op))
+      return true;
+    if (!op->use_empty())
+      return false;
+    if (auto stateOp = dyn_cast<StateOp>(op))
+      return stateOp.getLatency() == 0;
+    return false;
+  };
+  for (auto &op : *moduleOp.getBodyBlock())
+    if (isDead(&op))
+      erasureWorklist.insert(&op);
+  while (!erasureWorklist.empty()) {
+    auto *op = erasureWorklist.pop_back_val();
+    if (!isDead(op))
+      continue;
+    op->walk([&](Operation *innerOp) {
+      for (auto operand : innerOp->getOperands())
+        if (auto *defOp = operand.getDefiningOp())
+          if (!op->isProperAncestor(defOp))
+            erasureWorklist.insert(defOp);
+    });
+    op->erase();
```

Comment:
Shouldn't the operands of `op` also be added to the worklist?

Reply:
`op->walk` also visits `op` itself. Whenever I have to deal with nested ops, it feels like there must be some idiomatic MLIR way of getting all operands of an op and its nested subops. Instead I always end up writing this weird walk that tries to find operands that escape the confinese of the `op`.